File: packages/contracts/src/company-home.ts
import type {CanonicalResource} from "./index.js";
import type {AnalysisProjection} from "./semantic-analysis.js";
export interface CompanyHomeDescriptor {
 contract:"g8-company-home/1";
 company:CanonicalResource;company_label:string;valid_at:string;known_at:string;
 domain_packs:CanonicalResource[];analyses:AnalysisProjection[];
 financial_context:{authority:"ACCOUNTING_CONTEXT_ONLY";accounting_state:string;limitation:string;source_scopes:CanonicalResource[];ledgers:Array<{ledger:CanonicalResource;calendar:CanonicalResource|null;chart:CanonicalResource|null;currency:CanonicalResource|null;books:CanonicalResource[];periods:CanonicalResource[];context_ready:boolean}>};
 operations:{lens:"gas_network"|"enterprise_assets";authority:"GEOGRAPHY_CONTEXT_ONLY";valid_at:string;known_at:string;domain_pack_ids:string[];limitation:string};
 unavailable_financials:Array<{key:"profit_loss"|"balance_sheet"|"cash_flow"|"working_capital";label:string;reason:string}>;
 current_use_authorized:false;business_effect_authorized:false;
}

File: services/api/src/finai_api/domain/company_home.py
"""A company Home composes existing authorities; it does not create financial facts."""

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, model_validator

from finai_api.domain.resources import CanonicalResource
from finai_api.domain.semantic_analysis import Projection


class Model(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class CompanyHomeRequest(Model):
    company_id: UUID
    invocation_ids: list[UUID] = Field(default_factory=list, max_length=6)
    valid_at: AwareDatetime | None = None
    known_at: AwareDatetime | None = None

    @model_validator(mode="after")
    def distinct_results(self) -> "CompanyHomeRequest":
        if len(set(self.invocation_ids)) != len(self.invocation_ids):
            raise ValueError("Home selections must reference distinct retained invocations")
        return self


class HomeOperations(Model):
    lens: Literal["gas_network", "enterprise_assets"]
    authority: Literal["GEOGRAPHY_CONTEXT_ONLY"] = "GEOGRAPHY_CONTEXT_ONLY"
    valid_at: AwareDatetime
    known_at: AwareDatetime
    domain_pack_ids: list[UUID]
    limitation: str


class MissingFinancial(Model):
    key: Literal["profit_loss", "balance_sheet", "cash_flow", "working_capital"]
    label: str
    reason: str


class CompanyHomeDescriptor(Model):
    contract: Literal["g8-company-home/1"] = "g8-company-home/1"
    company: CanonicalResource
    company_label: str
    valid_at: AwareDatetime
    known_at: AwareDatetime
    domain_packs: list[CanonicalResource]
    analyses: list[Projection] = Field(max_length=6)
    operations: HomeOperations
    unavailable_financials: list[MissingFinancial]
    current_use_authorized: Literal[False] = False
    business_effect_authorized: Literal[False] = False
